dist/src/main/java/io/camunda/zeebe/shared/security/ProblemAuthFailureHandler.java

```diff
@@ -37,48 +38,57 @@ public ProblemAuthFailureHandler(final ObjectMapper objectMapper) {
   }
 
   @Override
-  public Mono<Void> onAuthenticationFailure(
-      final WebFilterExchange exchange, final AuthenticationException error) {
-    return handleFailure(exchange.getExchange(), HttpStatus.UNAUTHORIZED, error);
+  public void onAuthenticationFailure(
+      final HttpServletRequest request,
+      final HttpServletResponse response,
+      final AuthenticationException error)
+      throws IOException, ServletException {
+    handleFailure(request, response, HttpStatus.UNAUTHORIZED, error);
   }
 
   @Override
-  public Mono<Void> handle(final ServerWebExchange exchange, final AccessDeniedException error) {
+  public void handle(
+      final HttpServletRequest request,
+      final HttpServletResponse response,
+      final AccessDeniedException error)
+      throws IOException, ServletException {
     // if a token was passed but could not be validated, onAuthenticationFailure is called
     // however, if no token was passed, then access is denied here, and we want to distinguish
     // between unauthorized and forbidden; we can do that by checking the session principal to see
     // if it's authenticated or not
-    return exchange
-        .getPrincipal()
-        .flatMap(
-            principal -> {
-              if (principal instanceof final Authentication auth && auth.isAuthenticated()) {
-                return handleFailure(exchange, HttpStatus.FORBIDDEN, error);
-              }
+    final var principal = request.getUserPrincipal();
+    if (principal instanceof final Authentication auth && auth.isAuthenticated()) {
```

Comment:
❓ I dont fully get the logic here.

In the comments:

```
 however, if no token was passed, then access is denied here, and we want to distinguish
    // between unauthorized and forbidden; we can do that by checking the session principal to see
    // if it's authenticated or not
```

How can I pass no token but still be authenticated? 🤔

Reply:
I don't know why it says "no token" because if no token is passed, `#commence()` will be called to start the authentication flow. If a token is passed but is not valid, then `#onAuthenticationFailure()` would be called. Nevertheless, this `#handle()` method is supposed to be called when an authentication exists. Still, the user/client does not have the authority to execute the request (by throwing an access denied exception).

So it is unclear to me why it says "no token". @tmetzke, do you have any input?

Reply:
I have never seen this code before, sorry. It sounds a bit weird to me as well, yes. Do we know of any test case confirming this behavior and checking for `HttpStatus.FORBIDDEN`?

Reply:
I couldn't find any unit or integration test testing this part of the code.

In the context of this PR, I believe that's not crucial, as this PR just "migrates" the code from being reactive to non-reactive (using the Servlet stack). The "lifecycle" itself and its logical implementation remain the same. I am bit hesitant to change that right now.

Still would be good to understand it, but for now, I would just proceed 😄